File: web/domain/filing/models/income/capital-gains/capital-gains-securities-model.ts
/**
 * Capital Gains Securities Model - Wrapper for all securities capital gains types
 * Includes stocks, bonds, debentures, and mutual funds with aggregated totals
 */

import type { StocksCapitalGains } from './stocks-capital-gains-model'
import type { BondsCapitalGains } from './bonds-capital-gains-model'
import type { MutualFundsCapitalGains } from './mutual-funds-capital-gains-model'
import type { RsusCapitalGains } from './rsus-capital-gains-model'

/**
 * Master data for capital gains securities dropdowns (returned by API alongside data)
 */
export interface CapitalGainsMasterData {
  equitySharesTypes?: Array<{ value: string; label: string }>
  mutualFundEquityTypes?: Array<{ value: string; label: string }>
  rsuTypes?: Array<{ value: string; label: string }>
  debentureTypes?: Array<{ value: string; label: string }>
}

/**
 * Wrapper model for all securities capital gains types
 * Includes stocks, bonds, mutual funds, and RSUs with pre-calculated totals from backend.
 * Field names use camelCase to match the actual API JSON response.
 */
export interface CapitalGainsSecuritiesModel {
  filingId: number
  stocks?: StocksCapitalGains[] | null
  bonds?: BondsCapitalGains[] | null
  mutualFunds?: MutualFundsCapitalGains[] | null
  rsus?: RsusCapitalGains[] | null
  stocksCount?: number
  stocksTotal?: number
  bondsCount?: number
  bondsTotal?: number
  mutualFundsCount?: number
  mutualFundsTotal?: number
  rsusCount?: number
  rsusTotal?: number
  totalNetGain?: number
}

File: api/filing/models/capital_gains_securities_model.py
"""Capital gains securities models (stocks, bonds, mutual funds, RSUs)."""
from typing import List, Optional
from datetime import date
from .api_base_model import ApiBaseModel


class StocksCapitalGains(ApiBaseModel):
    """Model for stocks/equity shares capital gains."""
    stock_sale_id: int | None = None
    filing_id: int | None = None
    share_type: str | None = None
    date_of_sale: date | None = None
    date_of_purchase: date | None = None
    description_of_asset_sold: str | None = None
    total_sale_price: float = 0.0
    total_purchase_price: float = 0.0
    transfer_expenses: float | None = None
    stt_paid: bool | None = None
    fair_market_value: float | None = None
    net_gain: float | None = None
    broker: str | None = None


class BondsCapitalGains(ApiBaseModel):
    """Model for bonds and debentures capital gains."""
    bond_sale_id: int | None = None
    filing_id: int | None = None
    asset_type: str | None = None
    asset_description: str | None = None
    date_of_purchase: date | None = None
    date_of_sale: date | None = None
    total_purchase_price: float = 0.0
    total_sale_price: float = 0.0
    transfer_expenses: float | None = None
    net_gain: float | None = None
    broker: str | None = None

    @property
    def bond_type(self) -> str | None:
        return self.asset_type


class MutualFundsCapitalGains(ApiBaseModel):
    """Model for mutual funds capital gains."""
    mutual_fund_sale_id: int | None = None
    filing_id: int | None = None
    equity_type: str | None = None
    date_of_sale: date | None = None
    date_of_purchase: date | None = None
    description_of_asset_sold: str | None = None
    total_sale_price: float = 0.0
    total_purchase_price: float = 0.0
    transfer_expenses: float | None = None
    stt_paid: bool | None = None
    fair_market_value: float | None = None
    net_gain: float | None = None
    broker: str | None = None


class RsusCapitalGains(ApiBaseModel):
    """Model for RSU/Stock Options capital gains."""
    rsu_sale_id: int | None = None
    filing_id: int | None = None
    share_type: str | None = None
    description_of_asset_sold: str | None = None
    date_of_purchase: date | None = None
    date_of_sale: date | None = None
    total_purchase_price: float = 0.0
    total_sale_price: float = 0.0
    transfer_expenses: float | None = None
    fair_market_value: float | None = None
    net_gain: float | None = None
    broker: str | None = None
    stt_paid: bool | None = None


class CapitalGainsSecuritiesModel(ApiBaseModel):
    """Wrapper model for all securities capital gains types."""
    filing_id: int | None = None
    stocks: Optional[List[StocksCapitalGains]] = None
    bonds: Optional[List[BondsCapitalGains]] = None
    mutual_funds: Optional[List[MutualFundsCapitalGains]] = None
    rsus: Optional[List[RsusCapitalGains]] = None

    # Calculated totals
    stocks_count: int = 0
    stocks_total: float = 0.0
    bonds_count: int = 0
    bonds_total: float = 0.0
    mutual_funds_count: int = 0
    mutual_funds_total: float = 0.0
    rsus_count: int = 0
    rsus_total: float = 0.0
    total_net_gain: float = 0.0
